### ISSUE TITLE
Trigger tooltips on hover

### ISSUE THREAD

**Issue by steelbrain** [atom-community/linter#885]
That's how IDEs do it. and Atom is more than just an editor right?
I would love to accept a PR for this as I am mostly busy in personal work these days.

**Comment by steelbrain** [comment]
@blakeembrey What are your comments/suggestions on this one?

**Comment by blakeembrey** [comment]
@steelbrain Make sense. I was looking over the inline messenger PR earlier too. I don't have a lot of time to make this happen, unfortunately, but I can think about the experience a little. I think a few there's minor things that can add up together. 

At the very least, allowing hovering over the gutter might be useful - I also don't know how possible this is, but integrating with the scrollbar or minimap could be interesting.

**Comment by blakeembrey** [comment]
@steelbrain Actually, I was thinking some more, do you think we need a tooltip provider? If we're going to add tooltips, I feel like we should look at being a common service for all tooltips in the editor view to register to - for example, type information in TypeScript. That way we don't have conflicts between different peoples tooltips?

@basarat Thoughts, since you wrote the tooltip in the TypeScript package?

**Comment by basarat** [comment]
> do you think we need a tooltip provider

Definitely. It wasn't easy to get started ... but fortunately https://atom.io/packages/ide-haskell (@lierdakil) had solved that problem before me ... and I just copy pasted + converted to TypeScript. 


A sample of how to create a new *scope driven provider* might be https://github.com/atom-community/formatter (written in coffeescript) that brings in the pretty must *just the* scope detection code from autocomplete at some past point in time. Unless one is comfortable taking out the code from linter :)

**Comment by steelbrain** [comment]
@basarat Thank you for the examples and reference to the score driver provider. I am afraid we might not need any score-driven provider at all, as what we have is a text editor instance and some ranges. Have you looked at `inline-messenger` package yet? I had considered it in the past for this very purpose, we can all work on it together and add tooltips features to one package.

The codebase of it however wasn't as "neat" it looked like someone hacked out the inline tooltips from linter and implemented them in a separate package, but then again time fixes almost everything so that won't be a problem. so what do you suggest? Should we all collaborate on an existing package like that or build a new one with a solid API and strong foundations from the ground up?

**Comment by basarat** [comment]
I'm consumed by TSB https://github.com/basarat/tsb/tree/master beyond supporting what I've already started and keeping up apprearances at the moment. Sorry :rose:

**Comment by steelbrain** [comment]
@basarat The remote editing concept looks great, but it's gonna cost you a lot of time so best of luck with it :rose:

**Comment by basarat** [comment]
> The remote editing concept looks great, but it's gonna cost you a lot of time

contrary ... that was *very easy* ... Remote editing is what atom-typescript effectively does as well (parent->worker). I just added a lib to work over websockets and it was a go ;)

* Contract : https://github.com/basarat/tsb/blob/master/src/socket/socketContract.ts
* Client : https://github.com/basarat/tsb/blob/master/src/socket/socketClient.ts
* Server : https://github.com/basarat/tsb/blob/master/src/socket/socketServer.ts

:rose:

**Comment by steelbrain** [comment]
so we have two options:
  - Show tooltips when gutter is hovered (simpler but no IDE I know does this)
  - Show tooltips when text is hovered (complex but all IDEs I know do this)

We can do either or both of them, what are your suggestions?

**Comment by basarat** [comment]
> Show tooltips when text is hovered (complex but all IDEs I know do this)


Tricky : IDE's that do this e.g. Visual Studio annoy me because there are times when I want to see the *type information* (to figure out why the code is erroring in the first place) and they insist on showing me the error. 

As long as I still get type information on hover anything is fine :rose:

**Comment by jrial** [comment]
I don't have the expertise to help with this, so I'm not sure how much my opinion matters. But personally I think both would be nice (individually toggleable). Plus an option to disable the current tooltips when you're on the error, and instead simply display the error in the status bar.

**Comment by steelbrain** [comment]
@jrial We already have an option to disable the current tooltip, it's called `showInlineError` I think

**Comment by jrial** [comment]
@steelbrain yes, I saw that option. But what's missing is a way to still quickly see the error without opening the panel and hunting down the line number in the sometimes rather long list of issues. I kind of like the way Sublime does it, just putting the error message in the status bar, but I'd rather have the proposed hover over the gutter icon, TBH.

**Comment by steelbrain** [comment]
@jrial Sublime folks do not do it as a feature, they do it as a limitation. Also the status icon doesn't scale well, we have to make linter generic, a package that works for everybody, even people who have 5k+ error messages

**Comment by jrial** [comment]
@steelbrain yes, I already noticed that Sublime seems to only show one error per line. I guess that's indeed due to the space limitation in the status bar. When I said, "kind of like", I meant only in the sense that it's better than nothing.

I rather have what we have now. Or even better: what you're currently working on. My personal preference would go to the gutter, in addition to in the code, with the option to enable/disable both. But that's just my opinion, and I'm not contributing with code.

ps: you said you don't know any IDE that tooltips in the gutter? Two examples from personal experience would be Eclipse (http://goo.gl/efsyOC) and PyCharm (https://goo.gl/fgs97P).

**Comment by steelbrain** [comment]
@jrial I am a heavy JetBrains IDE user, I like how they have managed the minimap and embedded linter messages in it. We have a similar plugin too, it's called [`minimap-linter`](https://github.com/AtomLinter/atom-minimap-linter). IMO it doesn't make a difference how we show these errors, there was a difference when I first wrote this post, it was a performance difference but I have recently resolved it. The performance hit is now the same for both the gutter and the inline bubbles.

I have nothing against gutter decorations, I like them to be honest, but it's gonna be a big UI chance that doesn't give us any significant win. I would love to do that when we actually **need** that feature or it's requested by thousands of people :)
